src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AwaitExpressionSpiller.cs

```diff
@@ -673,19 +683,48 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
         {
             BoundSpillSequenceBuilder builder = null;
             var right = VisitExpression(ref builder, node.Right);
-            BoundExpression left;
-            if (builder == null || node.Left.Kind == BoundKind.Local)
+
+            BoundExpression left = node.Left;
+            if (builder == null)
             {
-                left = VisitExpression(ref builder, node.Left);
+                left = VisitExpression(ref builder, left);
             }
             else
             {
                 // if the right-hand-side has await, spill the left
                 var leftBuilder = new BoundSpillSequenceBuilder();
-                left = VisitExpression(ref leftBuilder, node.Left);
-                if (left.Kind != BoundKind.Local)
+
+                switch (left.Kind)
                 {
-                    left = Spill(leftBuilder, left, RefKind.Ref);
+                    case BoundKind.Local:
+                    case BoundKind.Parameter:
+                        // locals and parameters are directly assignable, LHS is not on the stack so nothing to spill
+                        break;
+
+                    case BoundKind.FieldAccess:
+                        var field = (BoundFieldAccess)left;
+                        // static fields are directly assignable, LHS is not on th estack, nothing to spill
+                        if (field.FieldSymbol.IsStatic) break;
+
+                        // instance fields are directly assignable, but receiver is pushed, so need to spill that.
+                        var receiver = VisitExpression(ref leftBuilder, field.ReceiverOpt);
+                        receiver = Spill(builder, receiver, field.FieldSymbol.ContainingType.IsValueType ? RefKind.Ref : RefKind.None);
+                        left = field.Update(receiver, field.FieldSymbol, field.ConstantValueOpt, field.ResultKind, field.Type);
+                        break;
+
+                    case BoundKind.ArrayAccess:
+                        var arrayAccess = (BoundArrayAccess)left;
+                        // array and indices are pushed on stack so need to spill that
+                        var expression = VisitExpression(ref leftBuilder, arrayAccess.Expression);
+                        expression = Spill(builder, expression, RefKind.None);
+                        var indices = this.VisitExpressionList(ref builder, arrayAccess.Indices, forceSpill: true);
+                        left = arrayAccess.Update(expression, indices, arrayAccess.Type);
+                        break;
+
+                    default:
+                        // must be indirectly assignable, just visit and spill
```

Comment:
The only case that we added for refs is when byref methods are passed _to_ "ref readonly" arguments, then we spill via a copy, since we do not need to make them assignable and do not care much about aliasing.

It would not work when the method is a target of an assignment where preserving assignability is a must. Those still demand that the left must be both decomposable (without losing sideeffects) and could be rehydrated into something that represents the same location.


---
In reply to: [127511548](https://github.com/dotnet/roslyn/pull/20883#discussion_r127511548) [](ancestors = 127511548,127502849)